### ISSUE TITLE
Harmony add not working

### ISSUE THREAD

**Issue by dmb0058** [bwssytems/ha-bridge#214]
I add the name, IP address, user name and password for my Harmony Hub on the Bridge Control page, click "Add". The correct details appear in habridge.config:

"harmonyaddress":{"devices":[{"name":"Harmony Hub", "ip":"192.168.0.156"}]}, "harmonyuser":"aaa@bbb.com", "harmonypwd":"xxxxxxxx"

I then press "Save" and the bridge re-initialises. Sometimes everything disappears:

"harmonyconfigured":false

Other times only the device disappears:

"harmonyuser":"aaa@bbb.com","harmonypwd":"xxxxxxxxxx"

Nothing else ever changes in any of these cases - no activities, devices etc.

**Comment by bwssytems** [comment]
What version are you running and on what system(RPi, Windows, Mac)?

**Comment by dmb0058** [comment]
Hi !

Sorry I should have said, it's 3.2.0 on a Pi.

All the best,

David

**Comment by bwssytems** [comment]
Great, now how are you staring the ha-bridge on the Pi

**Comment by dmb0058** [comment]
Hi,

I have Java 1.8 installed and am starting it from an init.d script with the
command "nohup java -jar /data/echobridge/ha-bridge-3.2.0.jar 2>&1 &".

I've been using it successfully to link an Echo Dot to my x.10 devices by
executing heyu commands on the Pi, but also have a Harmony Hub I'd like to
link in.

David

**Comment by bwssytems** [comment]
So, the question here is when you restart it does it have all the devices and other configuration? I am thinking this is running and the directory is not accessible. for the config and devices.

**Comment by dmb0058** [comment]
Hi,

Yes, it has all the devices. As I mentioned in the first post, the harmony
config details are written to the .config file when I press Add, but when
the bridge reinitialises they are deleted again, sometimes completely,
sometimes partially.

Cheers,

David

**Comment by bwssytems** [comment]
Ahhh, you need to hit save at the top.

**Comment by dmb0058** [comment]
Hi,

Yes, that seems to write the Harmony name/IP address/username/password to
the config file, and the file contents survive a reboot. However even when
I can see the details in the config file on the Pi, the values are not
visible on the "Bridge Control" page (the relevant fields are blank as if
no Harmony exists). Hitting the "Bridge Reinitialise" button causes the
details to be deleted from the config (I guess maybe this is what it's
supposed to do).

When the Harmony details are present in the config file the web application
looks the same - should I expect to see any new links or tabs ?

Also if I enter "http://192.168.0.104:8080/api/devices/harmony/activities"
in a web browser I get {"message":"A Harmony is not available."} back -
does this sound correct ? I'd expect the bridge to use the Harmony details
to interrogate the hub ...

Hope this helps,

David

**Comment by bwssytems** [comment]
Looks like you have a disconnect on where the ha-bridge is saving the data and it is not getting that info on start up. Please provide your startup configuration and how you start it and under what user.

**Comment by dmb0058** [comment]
See earlier notes for startup mode and command as root.

**Comment by dmb0058** [comment]
Hi,

I've installed all the files in /data/habridge on my Pi:

root@x10:/data/habridge# ls -l
total 21000
drwxr-xr-x 2 root root    4096 Nov  6 15:07 backups
-rw-r--r-- 1 root root    3095 Nov  2 08:05 device.db
-rw-r--r-- 1 root root 7160181 Sep  1 17:14 ha-bridge-3.1.0.jar
-rw-r--r-- 1 root root 7163557 Nov  2 08:40 ha-bridge-3.2.0.jar
-rw-r--r-- 1 root root 7163751 Nov  2 21:35 ha-bridge-3.2.1.jar
-rw------- 1 root root     667 Nov  6 14:56 habridge.config
root@x10:/data/habridge#

I then tried to run the bridge as root with this script:

#! /bin/sh
HABRIDGE_LOG=/var/log/heyu/ha-bridge.log
HABRIDGE_DIR=/data/habridge
cd $HABRIDGE_DIR
nohup java -jar $HABRIDGE_DIR/ha-bridge-3.2.0.jar
-Dconfig.file=$HABRIDGE_DIR/habridge.config > $HABRIDGE_LOG 2>&1 &

and I now get one error in the log file to say that the device.db file
can't be found, and one to say that some port (maybe 8080) is already in
use.

Any ideas welcome - log file attached.

Regards,

David

2016-11-06 14:59:06,748 [main] INFO  com.bwssystems.HABridge.HABridge - HA
Bridge (v3.2.0) starting....
2016-11-06 14:59:07,023 [main] INFO  com.bwssystems.HABridge.BridgeSettings
- reading from system properties
2016-11-06 14:59:07,111 [main] INFO  com.bwssystems.HABridge.BridgeSettings
- Adding 192.168.0.104 as our default upnp config address.
2016-11-06 14:59:07,118 [main] INFO  com.bwssystems.HABridge.HABridge - HA
Bridge (v3.2.0) initializing....
2016-11-06 14:59:07,372 [main] INFO  com.bwssystems.HABridge.SystemControl
- System control service started....
2016-11-06 14:59:09.601:INFO::Thread-0: Logging initialized @11212ms
2016-11-06 14:59:11,219 [main] WARN
 com.bwssystems.HABridge.dao.DeviceRepository - Error reading the file:
data/device.db - Does not exist or is not readable. continuing...
2016-11-06 14:59:11,228 [main] INFO
 com.bwssystems.HABridge.devicemanagmeent.DeviceResource - HABridge device
management service started....
2016-11-06 14:59:11,780 [main] INFO
 com.bwssystems.HABridge.upnp.UpnpSettingsResource - Hue description
service started....
2016-11-06 14:59:11,874 [main] INFO  com.bwssystems.util.UDPDatagramSender
- Initializing UDP response Seocket...
2016-11-06 14:59:11,975 [main] INFO  com.bwssystems.util.UDPDatagramSender
- UDP response Seocket initialized to: 50000
2016-11-06 14:59:12,746 [Thread-0] INFO  spark.webserver.JettySparkServer -
== Spark has ignited ...
2016-11-06 14:59:12,838 [Thread-0] INFO  spark.webserver.JettySparkServer -
>> Listening on 0.0.0.0:80
2016-11-06 14:59:12.991:INFO:oejs.Server:Thread-0: jetty-9.3.z-SNAPSHOT
2016-11-06 14:59:14,700 [Thread-0] ERROR spark.webserver.JettySparkServer -
ignite failed
java.net.BindException: Address already in use
    at sun.nio.ch.Net.bind0(Native Method)
    at sun.nio.ch.Net.bind(Net.java:414)
    at sun.nio.ch.Net.bind(Net.java:406)
    at
sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:214)
    at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
    at
org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:326)
    at
org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
    at
org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:244)
    at
org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
    at org.eclipse.jetty.server.Server.doStart(Server.java:384)
    at
org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
    at spark.webserver.JettySparkServer.ignite(JettySparkServer.java:131)
    at spark.SparkInstance.lambda$init$0(SparkInstance.java:341)
    at spark.SparkInstance$$Lambda$2/8801938.run(Unknown Source)
    at java.lang.Thread.run(Thread.java:744)

**Comment by audiofreak9** [comment]
@dmb0058 the device.db error appears on first run and always until the bridge runs successfully the first time.  Try adding this to your startup line 
`-Dserver.port=8080` 
OR whatever Port you choose, so it would be 
`nohup java -jar -Dserver.port=8080 $HABRIDGE_DIR/ha-bridge-3.2.0.jar
-Dconfig.file=$HABRIDGE_DIR/habridge.config > $HABRIDGE_LOG 2>&1 &`.  
Then to access your bridge go to http://&lt;Device IP&gt;:8080

**Comment by audiofreak9** [comment]
@dmb0058 BTW, it's best to auto-start the bridge with systemctl, follow the directions in this post here for that: http://forum.micasaverde.com/index.php/topic,31920.msg297253.html#msg297253

**Comment by dmb0058** [comment]
So the problem appears to be some mismatch between the files that are read at startup and those that are written in the course of running. I replaced the symbols (HABRIDGE_DIR/HABRIDGE_LOG) in my startup script with hard-coded paths and it's now working perfectly. I also found a directory "/etc/init.d/data" with a devices.db file - again suggests that the script was starting up in /etc/init.d, but then subsequent reads/writes went to the /data/habridge directory.

Thanks again - great piece of software.
